### PR TITLE
Rename transformer predecessors to references

### PR DESCRIPTION
**Pull request by MichaelBelousov**
Since 3.2.0 when the transformer started handling cyclic and backwards references, the term `predecessors` became inaccurate and confusing, since it implied that an element was definitely inserted before another, which is what the transformer used to only be able to handle (with some exceptions due to the old "deferral" mechanism). `getPredecessorIds`/`collectPredecessorIds` never really returned predecessors but references, and now we have `Element.requiredReferences` which returns real predecessors, adding to the confusion.

This PR deprecates (for removal in 4.x) `Element.getPredecessorIds` and `Element.collectPredecessorIds` and replaces them with `Element.getReferenceIds` and `Element.collectReferenceIds` to match the correct vocabulary. It also removes a lot of internal references to the "predecessor" vocabulary. 

This is a preliminary change to coming handling of backwards and cyclic references in aspects and link table relationship instances which the current lack of causes mistransformations in some models and transformations, especially some physical model consolidations with source provenance.

I have added a comment with a filtered diff below to help reviewers see what was changed that wasn't just a pure rename.

### CODE REVIEW

**Comment by MichaelBelousov** [comment]
If you subtract renames with some bash:
```sh
# 937f8d4b4da is the common commit with this branch and origin/master
 diff -u --color <(git diff --name-only 937f8d4b4da \
                        | xargs -i git show origin/master:{} \
                        | sed 's/predecessor/reference/g;s/Predecessor/Reference/g') \
                 <(git diff --name-only 937f8d4b4da | xargs -i git show HEAD:{})
```

you can get the remaining non-rename diff (as of 3cf9c1e7176197a48):

```diff
--- /dev/fd/63	2022-06-23 13:54:28.107465254 -0400
+++ /dev/fd/62	2022-06-23 13:54:28.107465254 -0400
@@ -1,4 +1,22 @@
-/*---------------------------------------------------------------------------------------------
+{
+  "changes": [
+    {
+      "packageName": "@itwin/core-backend",
+      "comment": "deprecate collectPredecessorIds and getPredecessorIds in favor of collectReferenceIds and getReferenceIds",
+      "type": "none"
+    }
+  ],
+  "packageName": "@itwin/core-backend"
+}{
+  "changes": [
+    {
+      "packageName": "@itwin/core-transformer",
+      "comment": "deprecate danglingPredecessorsBehavior in favor of danglingReferencesBehavior",
+      "type": "none"
+    }
+  ],
+  "packageName": "@itwin/core-transformer"
+}/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
@@ -876,10 +894,18 @@
   }
 
   /** Collect the Ids of this element's *references* at this level of the class hierarchy.
-   * A *reference* is an element that had to be inserted before this element could have been inserted.
+   * @deprecated use collectReferenceIds instead, the use of the term *predecessors* was confusing and became inaccurate when the transformer could handle cycles
+   * @beta
+   */
+  protected collectPredecessorIds(predecessorIds: Id64Set): void {
+    return this.collectReferenceIds(predecessorIds);
+  }
+
+  /** Collect the Ids of this entity's *references* at this level of the class hierarchy.
+   * A *reference* is any entity referenced by this entity's EC Data
    * This is important for cloning operations but can be useful in other situations as well.
    * @param referenceIds The Id64Set to populate with reference Ids.
-   * @note In order to clone/transform an element, all reference elements must have been previously cloned and remapped within the [IModelCloneContext]($backend).
+   * @note In order to clone/transform an entity, all referenced elements must have been previously cloned and remapped within the [IModelCloneContext]($backend).
    * @note This should be overridden (with `super` called) at each level the class hierarchy that introduces references.
    * @see getReferenceIds
    * @beta
@@ -892,7 +918,16 @@
       referenceIds.add(this.parent.id); // A parent element is a reference
   }
 
-  /** Get the Ids of this element's *references*. A *reference* is an element that had to be inserted before this element could have been inserted.
+  /** Get the Ids of this element's *references*. A *reference* is an element that this element references.
+   * This is important for cloning operations but can be useful in other situations as well.
+   * @beta
+   * @deprecated
+   */
+  public getPredecessorIds(): Id64Set {
+    return this.getReferenceIds();
+  }
+
+  /** Get the Ids of this element's *references*. A *reference* is an element that this element references.
    * This is important for cloning operations but can be useful in other situations as well.
    * @see collectReferenceIds
    * @beta
@@ -1003,6 +1038,7 @@
     referenceIds.add(this.category);
     // TODO: GeometryPartIds?
   }
+
   /** @beta */
   public static override readonly requiredReferenceKeys: ReadonlyArray<string> = [...super.requiredReferenceKeys, "category"];
 }
@@ -3514,6 +3550,21 @@
    *       reference depending on your use case.
    * @default "reject"
    * @beta
+   * @deprecated use danglingReferencesBehavior instead, the use of the term *predecessors* was confusing and became inaccurate when the transformer could handle cycles
+   */
+  danglingPredecessorsBehavior?: "reject" | "ignore";
+
+  /** The behavior to use when an element reference (id) is found stored as a reference on an element in the source,
+   * but the referenced element does not actually exist in the source.
+   * It is possible to craft an iModel with dangling references/invalidated relationships by, e.g., deleting certain
+   * elements without fixing up references.
+   *
+   * @note "reject" will throw an error and reject the transformation upon finding this case.
+   * @note "ignore" passes the issue down to consuming applications, iModels that have invalid element references
+   *       like this can cause errors, and you should consider adding custom logic in your transformer to remove the
+   *       reference depending on your use case.
+   * @default "reject"
+   * @beta
    */
   danglingReferencesBehavior?: "reject" | "ignore";
 
@@ -3676,7 +3727,8 @@
       // non-falsy defaults
       cloneUsingBinaryGeometry: options?.cloneUsingBinaryGeometry ?? true,
       targetScopeElementId: options?.targetScopeElementId ?? IModel.rootSubjectId,
-      danglingReferencesBehavior: options?.danglingReferencesBehavior ?? "reject",
+      // eslint-disable-next-line deprecation/deprecation
+      danglingReferencesBehavior: options?.danglingReferencesBehavior ?? options?.danglingPredecessorsBehavior ?? "reject",
     };
     this._isFirstSynchronization = this._options.wasSourceIModelCopiedToTarget ? true : undefined;
     // initialize exporter and sourceDb
@@ -3943,10 +3995,8 @@
 
   /** collect references this element has that are yet to be mapped, and if necessary create a
    * PartiallyCommittedElement for it to track resolution of unmapped references
-   * @returns {boolean}
    */
-  private collectUnmappedReferences(element: Element): boolean {
-
+  private collectUnmappedReferences(element: Element) {
     const missingReferences = new Set<string>();
     let thisPartialElem: PartiallyCommittedElement | undefined;
 
@@ -3988,8 +4038,6 @@
         this._pendingReferences.set({referenced: referenceId, referencer: element.id, isModelRef: false}, thisPartialElem);
       }
     }
-
-    return missingReferences.size > 0;
   }
 
   /** Cause the specified Element and its child Elements (if applicable) to be exported from the source iModel and imported into the target iModel.
@@ -4101,6 +4149,7 @@
       this.importer.importElement(targetElementProps); // don't need to import if iModel was copied
     }
     this.context.remapElement(sourceElement.id, targetElementProps.id!); // targetElementProps.id assigned by importElement
+
     // now that we've mapped this elem we can fix unmapped references to it
     for (const referencer of this._pendingReferences.getReferencers(sourceElement.id)) {
       const isModelRef = false; // we're in onExportElement so no
@@ -4110,6 +4159,7 @@
       pendingRef.resolveReference(sourceElement.id, isModelRef);
       this._pendingReferences.delete(key);
     }
+
     if (!this._options.noProvenance) {
       const aspectProps: ExternalSourceAspectProps = this.initElementProvenance(sourceElement.id, targetElementProps.id!);
       if (aspectProps.id === undefined) {
@@ -4762,7 +4812,7 @@
    * @param sourceTemplateModelId The Id of the template model in the sourceDb
    * @param targetModelId The Id of the target model (must be a subclass of GeometricModel3d) where the cloned component will be inserted.
    * @param placement The placement for the cloned component.
-   * @note *References* like the SpatialCategory must be remapped before calling this method.
+   * @note *Required References* like the SpatialCategory must be remapped before calling this method.
    * @returns The mapping of sourceElementIds from the template model to the instantiated targetElementIds in the targetDb in case further processing is required.
    */
   public async placeTemplate3d(sourceTemplateModelId: Id64String, targetModelId: Id64String, placement: Placement3d): Promise<Map<Id64String, Id64String>> {
@@ -4782,7 +4832,7 @@
    * @param sourceTemplateModelId The Id of the template model in the sourceDb
    * @param targetModelId The Id of the target model (must be a subclass of GeometricModel2d) where the cloned component will be inserted.
    * @param placement The placement for the cloned component.
-   * @note *References* like the DrawingCategory must be remapped before calling this method.
+   * @note *Required References* like the DrawingCategory must be remapped before calling this method.
    * @returns The mapping of sourceElementIds from the template model to the instantiated targetElementIds in the targetDb in case further processing is required.
    */
   public async placeTemplate2d(sourceTemplateModelId: Id64String, targetModelId: Id64String, placement: Placement2d): Promise<Map<Id64String, Id64String>> {
@@ -6384,8 +6434,7 @@
       geom: IModelTransformerTestUtils.createBox(Point3d.create(1, 1, 1)),
       placement: Placement3d.fromJSON({ origin: { x: 1 }, angles: {} }),
     } as PhysicalElementProps);
-    // because they are definition elements, display styles will be transformed first, but deferred until the excludedElements
-    // (which are references) are inserted
+    // because they are definition elements, display styles will be transformed first
     const myDisplayStyleId = DisplayStyle3d.insert(sourceDb, IModelDb.dictionaryId, "MyDisplayStyle3d", {
       excludedElements: [myPhysicalObjId],
     });
@@ -6609,8 +6658,7 @@
       geom: IModelTransformerTestUtils.createBox(Point3d.create(1, 1, 1)),
       placement: Placement3d.fromJSON({ origin: { x: 1 }, angles: {} }),
     } as PhysicalElementProps);
-    // because they are definition elements, display styles will be transformed first, but deferred until the excludedElements
-    // (which are references) are inserted
+    // because they are definition elements, display styles will be transformed first
     const myDisplayStyleId = DisplayStyle3d.insert(sourceDb, IModelDb.dictionaryId, "MyDisplayStyle3d", {
       excludedElements: [myPhysicalObjId],
     });
@@ -6728,6 +6776,8 @@
 
 The AuthStatus enum has been removed. This enum has fallen out of use since the authorization refactor in 3.0.0, and is no longer a member of [BentleyError]($core-bentley).
 
+The beta functions [Entity.collectPredecessorIds]($core-backend) and [Entity.getPredecessorIds]($core-backend) have been deprecated and replaced with [Entity.collectReferenceIds]($core-backend) and [Entity.getReferenceIds]($core-backend), since the term "predecessor" has been inaccurate since 3.2.0, when the transformer became capable of handling cyclic references and not just references to elements that were inserted before itself (predecessors).
+
 ### @itwin/core-mobile
 
 IOSApp, IOSAppOpts, and AndroidApp have been removed in favor of [MobileApp]($core-mobile) and [MobileAppOpts]($core-mobile). Developers were previously discouraged from making direct use of [MobileApp]($core-mobile), which was a base class of the two platform specific mobile apps. This distinction has been removed, as the implementation of the two apps was the same. IOSAppOpts, now [MobileAppOpts]($core-mobile), is an extension of [NativeAppOpts]($core-frontend) with the added condition that an [AuthorizationClient]($core-common) is never provided.
```

The bash gets much more verbose if you try make the diff show the file names

File: core/backend/src/Element.ts [hunk 1]
```diff
@@ -336,38 +336,55 @@ export class Element extends Entity {
...
+   * @deprecated
    */
   public getPredecessorIds(): Id64Set {
-    const predecessorIds = new Set<Id64String>();
-    this.collectPredecessorIds(predecessorIds);
-    return predecessorIds;
+    return this.getReferenceIds();
+  }
+
+  /** Get the Ids of this element's *references*. A *reference* is an element that this element references.
```
Comment (pmconne) [inline core/backend/src/Element.ts:373]: "A *reference* is an element that this element references." is rather tautological - can you be more descriptive?

Comment (MichaelBelousov) [reply core/backend/src/Element.ts:373]: you're right... I kept to the original desc format way too much, will fix

File: core/backend/src/Element.ts [hunk 2]
```diff
@@ -336,38 +336,55 @@ export class Element extends Entity {
...
 
-  /** Collect the Ids of this element's *predecessors* at this level of the class hierarchy.
-   * A *predecessor* is an element that had to be inserted before this element could have been inserted.
-   * This is important for cloning operations but can be useful in other situations as well.
-   * @param predecessorIds The Id64Set to populate with predecessor Ids.
-   * @note In order to clone/transform an element, all predecessor elements must have been previously cloned and remapped within the [IModelCloneContext]($backend).
-   * @note This should be overridden (with `super` called) at each level the class hierarchy that introduces predecessors.
-   * @see getPredecessorIds
+  /** Collect the Ids of this element's *references* at this level of the class hierarchy.
+   * @deprecated use collectReferenceIds instead, the use of the term *predecessors* was confusing and became inaccurate when the transformer could handle cycles
```
Comment (pmconne) [inline core/backend/src/Element.ts:340]: Can you use doc links when recommending alternatives for deprecated APIs (like [here](https://www.itwinjs.org/reference/appui-layout-react/toolbar/toolbarpanelalignment/))?

**Review by pmconne — APPROVED** [review]

File: core/backend/src/Element.ts (hunk 1 above)
Comment (MichaelBelousov) [reply core/backend/src/Element.ts:373]: how about now?

File: core/backend/src/Element.ts (hunk 2 above)
Comment (MichaelBelousov) [reply core/backend/src/Element.ts]: done for all 3 deprecations I made, please check

**Comment by pmconne** [comment]
@MichaelBelousov did you do an exhaustive grep? I just happened to notice a reference to `collectPredecessorIds` in RenderSchedule.ts in core/common.

**Comment by MichaelBelousov** [comment]
> @MichaelBelousov did you do an exhaustive grep? I just happened to notice a reference to `collectPredecessorIds` in RenderSchedule.ts in core/common.

I did but there were a lot of false positives for `predecessor` in the geometry stuff so I may have skipped some files accidentally while going through everything. I can make a second PR to fix that, I'll regrep

**Pull request merged**